pennylane_sf/gbs.py

```diff
@@ -0,0 +1,258 @@
+
+The Strawberry Fields variational GBS device is a simulator that provides a way to encode
+variational parameters into GBS so that the gradient with respect to the output probability
+distribution is accessible.
+
+Classes
+-------
+
+.. autosummary::
+   StrawberryFieldsGBS
+
+Code details
+~~~~~~~~~~~~
+"""
+from collections import OrderedDict
+
+import numpy as np
+import pennylane as qml
+import strawberryfields as sf
+from pennylane.operation import Probability
+from pennylane.wires import Wires
+from strawberryfields.ops import GraphEmbed, MeasureFock
+from strawberryfields.utils import all_fock_probs_pnr
+from thewalrus.quantum import find_scaling_adjacency_matrix as rescale
+from thewalrus.quantum import photon_number_mean_vector
+
+from .expectations import identity
+from .simulator import StrawberryFieldsSimulator
+
+
+class StrawberryFieldsGBS(StrawberryFieldsSimulator):
+    r"""StrawberryFields variational GBS device for PennyLane.
+
+    This device provides a simulator that can embed variational parameters into GBS such that the
+    analytic gradient of the probability distribution is accessible.
+
+    Args:
+        wires (int): the number of modes to initialize the device in
+        analytic (bool): indicates if the device should calculate expectations
+            and variances analytically
+        cutoff_dim (int): Fock-space truncation dimension
+        shots (int): Number of circuit evaluations/random samples used
+            to estimate expectation values of observables. If ``analytic=True``,
+            this setting is ignored.
+    """
+    name = "Strawberry Fields variational GBS PennyLane plugin"
+    short_name = "strawberryfields.gbs"
+
+    _operation_map = {
+        "ParamGraphEmbed": GraphEmbed,
+    }
+
+    _observable_map = {
+        "Identity": identity,
+    }
+
+    _circuits = {}
+
+    _capabilities = {"model": "cv", "provides_jacobian": True}
+
+    def __init__(self, wires, *, analytic=True, cutoff_dim, shots=1000):
+        super().__init__(wires, analytic=analytic, shots=shots)
+        self.cutoff = cutoff_dim
+        self._params = None
+        self._WAW = None
+
+    @staticmethod
+    def calculate_WAW(params, A, n_mean):
+        """Calculates the :math:`WAW` matrix.
+
+        Rescales :math:`A` so that when encoded in GBS the mean photon number is equal to
+        ``n_mean``.
+
+        Args:
+            params (array[float]): variable parameters
+            A (array[float]): adjacency matrix
+            n_mean (float): mean number of photons
+
+        Returns:
+            array[float]: the :math:`WAW` matrix
+        """
+        A *= rescale(A, n_mean)
+        W = np.diag(np.sqrt(params))
+        return W @ A @ W
+
+    @staticmethod
+    def calculate_n_mean(A):
+        """Calculates the mean number of photons for an adjacency matrix encoded into GBS.
+
+        Note that for ``A`` to be directly encoded into GBS, its singular values must not exceed
+        one.
```

Comment:
Nitpick:
```suggestion
        .. warning::
        
            Note that for ``A`` to be directly encoded into GBS, its singular values must
            not exceed one.
```

Reply:
Also, shouldn't this be 'its singular values must be less than 1'?

Another nitpick: This is likely something that could easily trip up users, since the only indication something went wrong would be a somewhat strange looking division by zero error. Is it worth having some validation to catch this, just to provide a more useful error message? Could even be as simple as

```python
singular_values = np.linalg.svd(A, compute_uv=False)

try:
    n_mean = np.sum(singular_values ** 2 / (1 - singular_values ** 2))
except DivideByZeroError:
    raise DeviceError("Singular values of matrix A must be less than 1")

return n_mean
```

Catching the divide by zero error is nice, as you don't have to worry about tolerances etc.